osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -32,14 +42,41 @@ protected DifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
         /// <returns>A structure describing the difficulty of the beatmap.</returns>
         public DifficultyAttributes Calculate(params Mod[] mods)
         {
-            mods = mods.Select(m => m.DeepClone()).ToArray();
+            preProcess(mods);
 
-            IBeatmap playableBeatmap = beatmap.GetPlayableBeatmap(ruleset.RulesetInfo, mods);
+            var skills = CreateSkills(Beatmap, playableMods, clockRate);
 
-            var track = new TrackVirtual(10000);
-            mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            if (!Beatmap.HitObjects.Any())
+                return CreateDifficultyAttributes(Beatmap, playableMods, skills, clockRate);
 
-            return calculate(playableBeatmap, mods, track.Rate);
+            foreach (var hitObject in getDifficultyHitObjects())
+            {
+                foreach (var skill in skills)
+                    skill.ProcessInternal(hitObject);
+            }
+
+            return CreateDifficultyAttributes(Beatmap, playableMods, skills, clockRate);
+        }
+
+        public IEnumerable<TimedDifficultyAttributes> CalculateTimed(params Mod[] mods)
```

Comment:
Is there a reason to expose this as a public live `IEnumerable`? Seems quite unsafe, are you okay with returning an array here instead?

Reply:
Sounds fine.

Reply:
I've changed this to a `List<>` as discussed. Please double-check.